### ISSUE TITLE
Validating fields in the context of others / uniqueness list issue

### ISSUE THREAD

**Issue by decentral1se** [pyeve/cerberus#467]
Used Cerberus version 1.2

---

- [x] I have the capacity to improve the docs when my problem is solved.
- [x] I have the capacity to submit a patch when a bug is identified.

----

#### Use-case abstract

Support validation of fields in the context of the entire schema.

#### Bug report / Feature request

In https://github.com/ansible/molecule/pull/1775, I needed to validate that the `name` key in the following schema has unique values. An example of the schema is:

```
platforms_docker_schema = {
    'platforms': {
        'type': 'list',
        'schema': {
            'type': 'dict',
            'schema': {
                'name': {
                    'type': 'string',
                },
                ...
    }
}
```

My solution involved adding a `'unique_keys': ['name'],` entry and then implementing a `_validate_unique_key`. However, there are two issues with this: 1) I cannot easily evaulate the name values in the context of the entire `platforms` name list (see https://github.com/ansible/molecule/pull/1775/files#diff-1e67bc3147c69bed463615849b56f3b1R948) 2) I cannot suppress duplicate uniqueness validation errors (see test in https://github.com/ansible/molecule/pull/1775/files#diff-ce21a0f23fce9df819e6b6241b30826aR39.

So, I would like to see if we can improve the API for my use case. Perhaps we can work on a 'list uniqueness' validator in core? Or some simpler solution?

Anyway, I'm new to Cerberus but here's my problem so far. 

Thanks for reading :+1: 

Some references that I found while searching:
  * https://github.com/pyeve/eve/issues/295
  * https://github.com/pyeve/cerberus/issues/60

**Comment by funkyfuture** [comment]
thank you for this inspiring issue.

first let me claim that your problem is solvable both with a custom rule or with a `check_with` test. i haven't looked at your code yet. let me first ask whether, beside the `document` property, you're aware of [these state-related attributes](http://docs.python-cerberus.org/en/stable/customize.html#relevant-validator-attributes). you may also look through the code how child validators can be employed (essentially it's three steps: derive a document and schema, pipe these through a child validator, remap the resulting errors into the error trees).

yes, a rule to check properties of collection containers seems reasonable. beside a `unique` constraint i can imagine a `sorted` would be commonly useful.

now, to assemble a schema with such rule and no need for a custom rule or `check_with` test, we'd have to introduce something like derivative fields to aggregate the items of interest, i your case, somthing like this:

```python
schema = {
    'platforms': {
        attributes("name"):  # this would map all items' 'name' attributes into a 'virtual' field
            {'collection': 'unique'}  # the new rule is then applied to the value of that ^
    }
}
```

it's complicated, so

1) please open separate issues for a new rule and such aggregation feature if you're interested in those.
2) especially the latter feature, i'd rather postpone this to the next major release. there would be a lot of code changes, some regressions due to that, aspects no one anticipated, but someone will tumble over.
3) if you want a solution soon, stick to a custom rule or `check_with` method.
4) maybe your use-case (w/ a rule or `check_with`) is better asked on StackOverflow in terms of (knowledge) sustainability.

**Comment by decentral1se** [comment]
> thank you for this inspiring issue.

:rainbow: 

Thanks for the thorough answer! I see now that it is indeed possible with the current API. I will make an attempt with `check_with` and also a child validator to see how it weighs up. StackOverflow probably was the better place for this after all ... :+1:

**Comment by decentral1se** [comment]
>  2\. i'd rather postpone this to the next major release. there would be a lot of code changes, some regressions due to that, aspects no one anticipated, but someone will tumble over.

Sounds scary!!! :)

**Comment by funkyfuture** [comment]
well, as new features usually originate from particular use-cases, other particular, yet unknown use-cases will require more adjustments down the road, but these will pop up after being released.

**Comment by decentral1se** [comment]
Ah, now I realise that `check_with` is only available in > 1.2 which isn't released.

I'm going to go with an ad-hoc solution for now (which gives duplicated errors).
